File: practice/leetcode/python/238_product_of_array_itself_self.py
#https://leetcode.com/problems/product-of-array-except-self/description/
from typing import List
import math

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        if nums.count(0) > 1: return [0] * len(nums)

        prod = math.prod(nums)

        product_without_zero = 1
        for num in nums:
            if num == 0: continue
            product_without_zero *= num

        answer = []
        for num in nums:
            if num == 0:
                answer.append(product_without_zero)
            else:
                answer.append(prod//num)

        return(answer)

File: practice/leetcode/cpp/238_product_of_array_itself_self.cpp
//https://leetcode.com/problems/product-of-array-except-self/description/
//
#include <vector>
using namespace std;

class Solution {
public:
	vector<int> productExceptSelt(vector<int> &nums) {
		int n = nums.size();
		vector<int> result(n, 1);
		//in first pass calculate the left multiplication
		for (int i = 1; i < n; i++) {
			result[i] = result[i-1]*nums[i-1];
		} 
		//in swcond pass calculate the right multiplication on the same array modified above
		int right = 1;
		for (int i = n-2; i >= 0; i--) {
			right *= nums[i+1];
			result[i] *= right;
		}
		return(result);

	}
};
